ceph/exporter.go

```diff
@@ -39,22 +39,36 @@ type Exporter struct {
 	RbdMirror bool
 	Logger    *logrus.Logger
 	Version   *Version
+	cc        []prometheus.Collector
 }
 
 // NewExporter returns an initialized *Exporter
 // We can choose to enable a collector to extract stats out of by adding it to the list of collectors.
 func NewExporter(conn Conn, cluster string, config string, user string, rgwMode int, logger *logrus.Logger) *Exporter {
-	return &Exporter{
+	e := &Exporter{
 		Conn:    conn,
 		Cluster: cluster,
 		Config:  config,
 		User:    user,
 		RgwMode: rgwMode,
 		Logger:  logger,
 	}
+	err := e.setCephVersion()
```

Comment:
Ah, so...I actually think this breaks the previous support for Version upgrades. The intended behaviour is that ceph_exporter automatically adapts to an upgrade, but now it will need a restart post-upgrade.

I think what we want here is for the collectors to not cache the version but to cache the exporter and read the version from there.

Reply:
Looking at the code I think the version's implementation sole purpose is to handle the different outputs between major ceph release https://github.com/digitalocean/ceph_exporter/blob/collectorInitFix/ceph/health.go#L561. It's passed down to the collectors for stuff like: https://github.com/digitalocean/ceph_exporter/blob/collectorInitFix/ceph/health.go#L994 so I think we should keep it

The reason why upgrade were supported is because we re-init all collectors all the time.

> but to cache the exporter and read the version from there.
I'm not sure I follow what you mean for that part, can you clarify? 

We could check for the version on every collection event and if needed re-init the collectors in order to keep the current behavior. That will reset stuff like `oldestInactivePGMap` tho.

 In all cases. the calls to version in `Collect()` and `Describe()` are still superfluous and should be removed I believe. 

I also overlooked the rbd-mirror stuff so I'll dig into this in a bit

Reply:
So, what I mean is that currently the collectors each cache the version within their own structs. That effectively freezes the version in the current code.

If instead they have a pointer to the exporter struct, they can query the version directly from there upon collection, and thus always get the latest, correct version.